ships/0027-image-abstraction-layer.md

```diff
@@ -0,0 +1,239 @@
+* The ability of mirroring remote images in an  “on-premise” image registry.
+
+## Motivation
+
+Keeping track of all Container Images in use in a Kubernetes cluster is a complicated task.
+Container Images may come from numerous different Image Registries. To add to this, Container
+Runtimes rely on remote registries (from the cluster's point of view) when obtaining Container
+Images, potentially making the process of pulling their blobs (manifests, config, and layers)
+slower.
+
+The notion of indexing Container Image versions by tags (“latest”, “development”, etc) is helpful.
+Still it does not provide users with the full confidence to always use the intended Container
+Image – today's "latest" tag might not be tomorrow's "latest" tag. In addition to that, these
+Image Registries allow access to Container Images by their Manifest content's hash (i.e., usually
+sha256), which gives users full confidence at a cost in semantics. Other factors may pop up when
+running, for instance, in an air-gapped environment, where the cluster nodes may not be allowed to
+reach any external Image Registry.
+
+This enhancement proposal suggests an image management abstraction layer. For instance, by
+providing a direct mapping between a Container Image tag (e.g., "latest") and its corresponding
+Manifest content's hash, users can then refer to the Container Image by its name – and be sure to
+use that specific version.
+
+In summary, the proposed operator mirrors and indexes remote Container Images into a Kubernetes
+cluster, allowing for other Operators to easily navigate through these multiple indexed Images.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Summary
+
+Design the Kubernetes API types (CRDs) and create a minimal viable product for such an abstraction
+layer.
+
+### Non-Goals
+
+At this stage to provide the integration with Shipwright’ builds is not a goal of this proposal,
+this will need to be tackled once we have the basic types and functionality in place. To provide
+Image storage for Container Images in the form of an integrated Image Registry inside the cluster
+is also not part of the scope of this proposal.
+
+## Proposal
+
+The operator would leverage a _custom resource definition_ called Image. An Image represents an
+image tag in a remote registry. For instance, an Image called `myapp-devel` may be created to keep
+track of the image `quay.io/company/myapp:latest`. A Image _custom resource_ layout would look
+like this:
+
+```yaml
+apiVersion: shipwright.io/v1alpha1
+kind: Image
+metadata:
+  name: myapp-devel
+  namespace: my-namespace
+spec:
+  source: quay.io/company/myapp:latest
+  mirror: false
```

Comment:
Something to consider for the future - having the ability to specify mirrors on a per-image basis.

The current design implies that there is a single mirror registry that is centrally configured. Having the ability to specify the mirror configuration directly for an image (host, auth, etc.) may be useful.